### ISSUE TITLE
require/typed doesn't protect with contract properly

### ISSUE THREAD

**Issue by sorawee** [racket/typed-racket#1256]
### What version of Racket are you using?

8.5 [cs]

### What program did you run?

```
#lang racket

(module a typed/racket
  (provide f)
  (: f (-> Byte Integer))
  (define (f x)
    (add1 x)))

(module b typed/racket
  (require/typed (submod ".." a)
                 [f (-> Byte Integer)])
  (provide f))

(require 'b)

(f "a")
```

### What should have happened?

A contract error, saying that `"a"` is not a number.

### If you got an error message, please include it here.

```
f: broke its own contract
  promised: exact-integer?
  produced: #<garbage>
  in: (-> any/c exact-integer?)
  contract from: (interface for f)
  blaming: (interface for f)
   (assuming the contract is correct)
```

In DrRacket, running the program repeatedly and quickly will crash DrRacket.

**Comment by capfredf** [comment]
> In DrRacket, running the program repeatedly and quickly will crash DrRacket.

This is surprising.  cc @rfindler

**Comment by capfredf** [comment]
@sorawee I can't reproduce a crash on Linux with 8.5 cs

**Comment by samth** [comment]
The `#<garbage>` suggests that we're causing memory unsafety, so DrRacket crashing is Typed Racket's fault.

**Comment by samth** [comment]
I also get `#<garbage>` in the error on my HEAD linux version, with CS, so this bug is definitely real.

**Comment by capfredf** [comment]
Yeah, I get `#<garbage>` as well. I just thought DrRacket had some kind of isolation between it and programs it ran

**Comment by samth** [comment]
Ok, I think this is a simple miscommunication between the modules. `a` thinks that it's providing `f` to a typed module, so it doesn't need a contract. `b` thinks that it doesn't need to check the domain of `f` in `require/typed` because `b` is typed, and that it doesn't need to add protection to `f` when providing it because it doesn't define it locally. 

Each of those decisions is sensible locally, but they don't all work together.

**Comment by bennn** [comment]
So, should `b` add protection when providing?

In general: `require/typed` imports from a typed module get full contracts.

**Comment by rfindler** [comment]
It does, but bugs like this are defeating that isolation. The isolation is
in the language, not at the OS layer.

**Comment by samth** [comment]
@bennn Yes, I agree (basically we should just make `require/typed` not be a typed context).
